iree/compiler/Codegen/SPIRV/SPIRVExpandIfRegion.cpp

```diff
@@ -0,0 +1,257 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===----------------------------------------------------------------------===//
+//
+// This file implements patterns and passes for expanding scf.if's regions
+// by pulling in ops before and after the scf.if op into both regions of the
+// scf.if op.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "llvm/ADT/SetVector.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Dialect/Affine/Analysis/AffineStructures.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/SCF/Passes.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/SCF/Transforms.h"
+#include "mlir/Dialect/SCF/Utils/AffineCanonicalizationUtils.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/RegionUtils.h"
+
+#define DEBUG_TYPE "iree-spirv-expand-if-region"
+
+// Note: A pull request is open to upstream this pattern:
+//   https://reviews.llvm.org/D117019
+// After it lands, this pattern can be replaced.
+
+namespace mlir {
+namespace iree_compiler {
+
+/// Pulls ops at the same nest level as the given `ifOp` into both regions of
```

Comment:
Have discussed this before, but this is a strange pattern to me. IIUC, this is pulling in ops from after the `if` into both the `if` and `else` parts to trigger some canonicalizations. That seems a bit backwards, and might be some canonicalizations through interface that is missing. Just a note to say I feel like there are potential footguns here.

Reply:
The purpose of creating the `scf.if` op and then pulling all ops before and after into both regions is to create the fast and slow path, where in the former we handle inner tiles so everything is nice, in the later we need to have conditional reads. This segregation helps for performance given that we don't want to have `scf.if` in deep loop nests as much as possible. So it's not just trying to put ops together to make canonicalization easy. (There is indeed an aspect of that too though.)